// @slonik/sql-tag/dist/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnnestToken = exports.TimestampToken = exports.QueryToken = exports.ListToken = exports.JsonToken = exports.JsonBinaryToken = exports.IntervalToken = exports.IdentifierToken = exports.FragmentToken = exports.DateToken = exports.ComparisonPredicateToken = exports.BinaryToken = exports.ArrayToken = void 0;
exports.ArrayToken = Symbol.for('SLONIK_TOKEN_ARRAY');
exports.BinaryToken = Symbol.for('SLONIK_TOKEN_BINARY');
exports.ComparisonPredicateToken = Symbol.for('SLONIK_TOKEN_COMPARISON_PREDICATE');
exports.DateToken = Symbol.for('SLONIK_TOKEN_DATE');
exports.FragmentToken = Symbol.for('SLONIK_TOKEN_FRAGMENT');
exports.IdentifierToken = Symbol.for('SLONIK_TOKEN_IDENTIFIER');
exports.IntervalToken = Symbol.for('SLONIK_TOKEN_INTERVAL');
exports.JsonBinaryToken = Symbol.for('SLONIK_TOKEN_JSON_BINARY');
exports.JsonToken = Symbol.for('SLONIK_TOKEN_JSON');
exports.ListToken = Symbol.for('SLONIK_TOKEN_LIST');
exports.QueryToken = Symbol.for('SLONIK_TOKEN_QUERY');
exports.TimestampToken = Symbol.for('SLONIK_TOKEN_TIMESTAMP');
exports.UnnestToken = Symbol.for('SLONIK_TOKEN_UNNEST');
