// slonik/dist/factories/createConnectionPool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConnectionPool = void 0;
const Logger_1 = require("../Logger");
const utilities_1 = require("@slonik/utilities");
const promises_1 = require("node:timers/promises");
const serialize_error_1 = require("serialize-error");
const logger = Logger_1.Logger.child({
    namespace: 'createConnectionPool',
});
const createConnectionPool = ({ driver, poolSize = 1, }) => {
    // See test "waits for all connections to be established before attempting to terminate the pool"
    // for explanation of why `pendingConnections` is needed.
    const pendingConnections = [];
    const connections = [];
    const waitingClients = [];
    const id = (0, utilities_1.generateUid)();
    let isEnding = false;
    let isEnded = false;
    let poolEndPromise = null;
    const endPool = async () => {
        try {
            await Promise.all(pendingConnections);
        }
        catch (error) {
            logger.error({
                error: (0, serialize_error_1.serializeError)(error),
            }, 'error in pool termination sequence while waiting for pending connections to be established');
        }
        try {
            await Promise.all(waitingClients.map((waitingClient) => waitingClient.deferred.promise));
        }
        catch (error) {
            logger.error({
                error: (0, serialize_error_1.serializeError)(error),
            }, 'error in pool termination sequence while waiting for waiting clients to be resolved');
        }
        // This is needed to ensure that all pending connections were assigned a waiting client.
        // e.g. "waits for all connections to be established before attempting to terminate the pool" test
        await (0, promises_1.setTimeout)(0);
        await Promise.all(connections.map((connection) => connection.destroy()));
    };
    const acquire = async () => {
        if (isEnding) {
            throw new Error('Connection pool is being terminated.');
        }
        if (isEnded) {
            throw new Error('Connection pool has ended.');
        }
        const idleConnection = connections.find((connection) => connection.state() === 'IDLE');
        if (idleConnection) {
            idleConnection.acquire();
            return idleConnection;
        }
        if (pendingConnections.length + connections.length < poolSize) {
            const pendingConnection = driver.createClient();
            pendingConnections.push(pendingConnection);
            const connection = await pendingConnection;
            const onRelease = () => {
                const waitingClient = waitingClients.shift();
                if (!waitingClient) {
                    return;
                }
                if (connection.state() !== 'IDLE') {
                    throw new Error('Connection is not idle.');
                }
                connection.acquire();
                waitingClient.deferred.resolve(connection);
            };
            connection.on('release', onRelease);
            const onDestroy = () => {
                connection.removeListener('release', onRelease);
                connection.removeListener('destroy', onDestroy);
                connections.splice(connections.indexOf(connection), 1);
                const waitingClient = waitingClients.shift();
                if (!waitingClient) {
                    return;
                }
                // eslint-disable-next-line promise/prefer-await-to-then
                acquire().then(waitingClient.deferred.resolve, waitingClient.deferred.reject);
            };
            connection.on('destroy', onDestroy);
            connection.acquire();
            connections.push(connection);
            pendingConnections.splice(pendingConnections.indexOf(pendingConnection), 1);
            return connection;
        }
        else {
            const deferred = (0, utilities_1.defer)();
            waitingClients.push({
                deferred,
            });
            const queuedAt = process.hrtime.bigint();
            logger.warn({
                connections: connections.length,
                pendingConnections: pendingConnections.length,
                poolSize,
                waitingClients: waitingClients.length,
            }, `connection pool full; client has been queued`);
            // eslint-disable-next-line promise/prefer-await-to-then
            return deferred.promise.then((connection) => {
                logger.debug({
                    connectionId: connection.id(),
                    duration: Number(process.hrtime.bigint() - queuedAt) / 1e6,
                }, 'connection has been acquired from the queue');
                return connection;
            });
        }
    };
    return {
        acquire,
        end: async () => {
            isEnding = true;
            if (poolEndPromise) {
                return poolEndPromise;
            }
            poolEndPromise = endPool();
            isEnded = true;
            return poolEndPromise;
        },
        id: () => {
            return id;
        },
        state: () => {
            const stateName = isEnded ? 'ENDED' : isEnding ? 'ENDING' : 'ACTIVE';
            const state = {
                acquiredConnections: 0,
                idleConnections: 0,
                pendingDestroyConnections: 0,
                pendingReleaseConnections: 0,
            };
            // TODO add pendingAcquireConnections
            // TODO add destroyedConnections
            for (const connection of connections) {
                if (connection.state() === 'ACQUIRED') {
                    state.acquiredConnections++;
                }
                if (connection.state() === 'IDLE') {
                    state.idleConnections++;
                }
                if (connection.state() === 'PENDING_RELEASE') {
                    state.pendingReleaseConnections++;
                }
                if (connection.state() === 'PENDING_DESTROY') {
                    state.pendingDestroyConnections++;
                }
                if (connection.state() === 'DESTROYED') {
                    state.pendingReleaseConnections++;
                }
            }
            return {
                ...state,
                state: stateName,
                waitingClients: waitingClients.length,
            };
        },
    };
};
exports.createConnectionPool = createConnectionPool;
