// slonik/dist/routines/executeQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeQuery = void 0;
const constants_1 = require("../constants");
const transactionContext_1 = require("../contexts/transactionContext");
const state_1 = require("../state");
const errors_1 = require("@slonik/errors");
const utilities_1 = require("@slonik/utilities");
const get_stack_trace_1 = require("get-stack-trace");
const serialize_error_1 = require("serialize-error");
const retryQuery = async (connectionLogger, connection, query, retryLimit) => {
    let result;
    let remainingRetries = retryLimit;
    let attempt = 0;
    // @todo Provide information about the queries being retried to the logger.
    while (remainingRetries-- > 0) {
        attempt++;
        try {
            connectionLogger.trace({
                attempt,
                queryId: query.executionContext.queryId,
            }, 'retrying query');
            result = await query.executionRoutine(connection, query.sql, 
            // @todo Refresh execution context to reflect that the query has been re-tried.
            query.values, 
            // This (probably) requires changing `queryId` and `queryInputTime`.
            // It should be needed only for the last query (because other queries will not be processed by the middlewares).
            query.executionContext, {
                sql: query.sql,
                values: query.values,
            });
            // If the attempt succeeded break out of the loop
            break;
        }
        catch (error) {
            if (typeof error.code === 'string' &&
                error.code.startsWith(constants_1.TRANSACTION_ROLLBACK_ERROR_PREFIX) &&
                remainingRetries > 0) {
                continue;
            }
            throw error;
        }
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return result;
};
// eslint-disable-next-line complexity
const executeQuery = async (connectionLogger, connection, clientConfiguration, query, inheritedQueryId, executionRoutine, stream = false) => {
    const poolClientState = (0, state_1.getPoolClientState)(connection);
    if (poolClientState.terminated) {
        throw new errors_1.BackendTerminatedError(poolClientState.terminated);
    }
    if (query.sql.trim() === '') {
        throw new errors_1.InvalidInputError('Unexpected SQL input. Query cannot be empty.');
    }
    if (query.sql.trim() === '$1') {
        throw new errors_1.InvalidInputError('Unexpected SQL input. Query cannot be empty. Found only value binding.');
    }
    const transactionStore = transactionContext_1.transactionContext.getStore();
    if (clientConfiguration.dangerouslyAllowForeignConnections !== true &&
        transactionStore?.transactionId &&
        transactionStore.transactionId !== poolClientState.transactionId) {
        throw new errors_1.UnexpectedForeignConnectionError();
    }
    const queryInputTime = process.hrtime.bigint();
    let stackTrace = null;
    if (clientConfiguration.captureStackTrace) {
        stackTrace = (0, get_stack_trace_1.getStackTrace)();
    }
    const queryId = inheritedQueryId ?? (0, utilities_1.generateUid)();
    const log = connectionLogger.child({
        queryId,
    });
    const originalQuery = {
        // See comments in `formatSlonikPlaceholder` for more information.
        sql: query.sql.replaceAll('$slonik_', '$'),
        values: query.values,
    };
    let actualQuery = {
        ...originalQuery,
    };
    const executionContext = {
        connectionId: poolClientState.connectionId,
        log,
        originalQuery,
        poolId: poolClientState.poolId,
        queryId,
        queryInputTime,
        resultParser: query.parser,
        sandbox: {},
        stackTrace,
        transactionId: poolClientState.transactionId,
    };
    for (const interceptor of clientConfiguration.interceptors) {
        if (interceptor.beforeTransformQuery) {
            await interceptor.beforeTransformQuery(executionContext, actualQuery);
        }
    }
    for (const interceptor of clientConfiguration.interceptors) {
        if (interceptor.transformQuery) {
            actualQuery = interceptor.transformQuery(executionContext, actualQuery);
        }
    }
    let result;
    if (!stream) {
        for (const interceptor of clientConfiguration.interceptors) {
            if (interceptor.beforeQueryExecution) {
                result = await interceptor.beforeQueryExecution(executionContext, actualQuery);
                if (result) {
                    log.info('beforeQueryExecution interceptor produced a result; short-circuiting query execution using beforeQueryExecution result');
                    return result;
                }
            }
        }
    }
    const notices = [];
    const noticeListener = (notice) => {
        notices.push(notice);
    };
    const activeQuery = (0, utilities_1.defer)();
    const blockingPromise = poolClientState.activeQuery?.promise ?? null;
    poolClientState.activeQuery = activeQuery;
    await blockingPromise;
    connection.on('notice', noticeListener);
    const queryWithContext = {
        executionContext,
        executionRoutine,
        sql: actualQuery.sql,
        values: actualQuery.values,
    };
    try {
        try {
            try {
                result = await executionRoutine(connection, actualQuery.sql, actualQuery.values, executionContext, actualQuery);
            }
            catch (error) {
                const shouldRetry = typeof error.code === 'string' &&
                    error.code.startsWith(constants_1.TRANSACTION_ROLLBACK_ERROR_PREFIX) &&
                    clientConfiguration.queryRetryLimit > 0;
                // Transactions errors in queries that are part of a transaction are handled by the transaction/nestedTransaction functions
                if (shouldRetry && !poolClientState.transactionId) {
                    result = await retryQuery(connectionLogger, connection, queryWithContext, clientConfiguration.queryRetryLimit);
                }
                else {
                    throw error;
                }
            }
        }
        catch (error) {
            // The driver is responsible for throwing an appropriately wrapped error.
            if (error instanceof errors_1.BackendTerminatedError) {
                poolClientState.terminated = error;
            }
            // If the error has been already handled by the driver, then we should not wrap it again.
            if (!(error instanceof errors_1.SlonikError)) {
                if (error.message === 'Connection terminated unexpectedly') {
                    throw new errors_1.BackendTerminatedUnexpectedlyError(error);
                }
                if (error.message.includes('tuple to be locked was already moved to another partition due to concurrent update')) {
                    throw new errors_1.TupleMovedToAnotherPartitionError(error);
                }
            }
            error.notices = notices;
            throw error;
        }
        finally {
            connection.off('notice', noticeListener);
            activeQuery.resolve(null);
        }
    }
    catch (error) {
        log.error({
            error: (0, serialize_error_1.serializeError)(error),
        }, 'execution routine produced an error');
        for (const interceptor of clientConfiguration.interceptors) {
            if (interceptor.queryExecutionError) {
                await interceptor.queryExecutionError(executionContext, actualQuery, error, notices);
            }
        }
        error.notices = notices;
        throw error;
    }
    if (!result) {
        throw new errors_1.UnexpectedStateError('Expected query result to be returned.');
    }
    // @ts-expect-error -- We want to keep notices as readonly for consumer, but write to it here.
    result.notices = notices;
    if (result.type === 'QueryResult') {
        for (const interceptor of clientConfiguration.interceptors) {
            if (interceptor.afterQueryExecution) {
                await interceptor.afterQueryExecution(executionContext, actualQuery, result);
            }
        }
        const interceptors = clientConfiguration.interceptors.slice();
        for (const interceptor of interceptors) {
            if (interceptor.transformRow) {
                const { transformRow } = interceptor;
                const { fields } = result;
                const rows = await Promise.all(result.rows.map((row) => transformRow(executionContext, actualQuery, row, fields)));
                result = {
                    ...result,
                    rows,
                };
            }
        }
        for (const interceptor of clientConfiguration.interceptors) {
            if (interceptor.beforeQueryResult) {
                await interceptor.beforeQueryResult(executionContext, actualQuery, result);
            }
        }
    }
    return result;
};
exports.executeQuery = executeQuery;
