// @slonik/sql-tag/dist/sqlFragmentFactories/createIntervalSqlFragment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createIntervalSqlFragment = void 0;
const tokens_1 = require("../tokens");
const formatSlonikPlaceholder_1 = require("../utilities/formatSlonikPlaceholder");
const errors_1 = require("@slonik/errors");
const zod_1 = require("zod");
const IntervalInput = zod_1.z
    .object({
    days: zod_1.z.number().optional(),
    hours: zod_1.z.number().optional(),
    minutes: zod_1.z.number().optional(),
    months: zod_1.z.number().optional(),
    seconds: zod_1.z.number().optional(),
    weeks: zod_1.z.number().optional(),
    years: zod_1.z.number().optional(),
})
    .strict();
const intervalFragments = [
    'years',
    'months',
    'weeks',
    'days',
    'hours',
    'minutes',
    'seconds',
];
const tokenMap = {
    minutes: 'mins',
    seconds: 'secs',
};
const createIntervalSqlFragment = (token, greatestParameterPosition) => {
    let intervalInput;
    try {
        intervalInput = IntervalInput.parse(token.interval);
    }
    catch {
        throw new errors_1.InvalidInputError('Interval input must not contain unknown properties.');
    }
    const values = [];
    const intervalTokens = [];
    for (const intervalFragment of intervalFragments) {
        const value = intervalInput[intervalFragment];
        if (value !== undefined) {
            values.push(value);
            const mappedToken = tokenMap[intervalFragment] ?? intervalFragment;
            intervalTokens.push(mappedToken +
                ' => ' +
                (0, formatSlonikPlaceholder_1.formatSlonikPlaceholder)(greatestParameterPosition + values.length));
        }
    }
    return {
        sql: 'make_interval(' + intervalTokens.join(', ') + ')',
        type: tokens_1.FragmentToken,
        values,
    };
};
exports.createIntervalSqlFragment = createIntervalSqlFragment;
